src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListenerBase.cs

```diff
@@ -272,22 +286,123 @@ private static async Task ProcessWorkCoreAsync(ImmutableArray<Work> work, Stream
+    private async Task ReportUpdateProjectAsync(Stream stream, Project project, CancellationToken cancellationToken)
     {
-        logger.LogTrace("Serializing information for {projectId}", project.Id);
-        var projectInfo = await RazorProjectInfoFactory.ConvertAsync(project, logger, cancellationToken).ConfigureAwait(false);
-        if (projectInfo is null)
+        _logger.LogTrace("Serializing information for {projectId}", project.Id);
+        var projectPath = Path.GetDirectoryName(project.FilePath);
+        if (projectPath is null)
         {
-            logger.LogTrace("Skipped writing data for {projectId}", project.Id);
+            _logger.LogInformation("projectPath is null, skip update for {projectId}", project.Id);
             return;
         }
 
-        stream.WriteProjectInfoAction(ProjectInfoAction.Update);
-        await stream.WriteProjectInfoAsync(projectInfo, cancellationToken).ConfigureAwait(false);
+        var intermediateOutputPath = Path.GetDirectoryName(project.CompilationOutputInfo.AssemblyPath);
+        if (intermediateOutputPath is null)
+        {
+            _logger.LogInformation("intermediateOutputPath is null, skip update for {projectId}", project.Id);
+            return;
+        }
+
+        if (!_projectEntryMap.TryGetValue(project.Id, out var entry))
+        {
+            _projectEntryMap[project.Id] = entry = new();
+        }
+
+        var (didChange, projectInfo) = await DidProjectInfoChangeAsync(project, entry, projectPath, intermediateOutputPath, cancellationToken).ConfigureAwait(false);
+        if (didChange)
+        {
+            projectInfo.AssumeNotNull();
+            UpdateEntry(project, projectInfo, entry);
+
+            stream.WriteProjectInfoAction(ProjectInfoAction.Update);
+            await stream.WriteProjectInfoAsync(projectInfo, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    private void UpdateEntry(Project project, RazorProjectInfo projectInfo, ProjectEntry entry)
+    {
+        entry.RootNamespace = projectInfo.RootNamespace;
+        entry.ConfigurationHash = projectInfo.Configuration.GetHashCode();
+        entry.DocumentsHash = projectInfo.Documents.GetHashCode();
+
+        _cachedTagHelperResolver.TryGetId(project.Id, out var resultId);
+        entry.TagHelpersResultId = resultId;
+    }
+
+    private async Task<(bool, RazorProjectInfo? projectInfo)> DidProjectInfoChangeAsync(Project project, ProjectEntry entry, string projectPath, string intermediateOutputPath, CancellationToken cancellationToken)
+    {
+        ImmutableArray<DocumentSnapshotHandle> documents = default;
+        var csharpLanguageVersion = (project.ParseOptions as CSharpParseOptions)?.LanguageVersion ?? LanguageVersion.Default;
+        CachedTagHelperResolver.DeltaResult? delta = null;
+
+        var configuration = RazorProjectInfoHelpers.ComputeRazorConfigurationOptions(project.AnalyzerOptions.AnalyzerConfigOptionsProvider, out var rootNamespace);
```

Comment:
Consider returning a tuple rather than one return value and one out parameter.

Reply:
Given the null check, `bool Try...(out, out)` could also be appropriate

Reply:
It actually can't be null, but I can make this a tuple. The null check isn't needed